src/metax_api/middleware/identifyapicaller.py

```diff
@@ -0,0 +1,128 @@
+from base64 import b64decode
+import yaml
+
+from django.conf import settings as django_settings
+from django.http import HttpResponseForbidden
+from metax_api.exceptions import Http403
+from metax_api.utils import executing_test_case, executing_travis
+
+import logging
+_logger = logging.getLogger(__name__)
+d = logging.getLogger(__name__).debug
+
+"""
+Currently API access authnz is already enforced once on the web server proxy level.
+For fine-grained authorization per action, or user-specific special logic, this
+middleware stores the API user information to request.user.username, which will
+be accessible in views.
+
+A middleware is initialized only once during the lifetime of a process, and users
+are loaded during middleware initialization. Thus when new api users are added to
+app_config, the processes need to be restarted. This should be a rare enough occasion
+to not be a cataclysmic event for the project.
+
+The object request.user is Django's default AnonymousUser-object, since we are not using
+ordinary Django authz methods, nor are we storing any sessions. Meaning user authnz
+is performed on each request. This, however, is very fast, since there are no db calls to
+fetch users or sessions: Everything required for user authnz is living inside the middleware object
+for the lifetime of the process.
+"""
+
+def IdentifyApiCaller(*args, **kwargs):
+    if executing_test_case() or executing_travis():
+        return _IdentifyApiCallerDummy(*args, **kwargs)
+    else:
+        return _IdentifyApiCaller(*args, **kwargs)
+
+
+class _IdentifyApiCaller():
+
+    def __init__(self, get_response):
+        self.get_response = get_response
+        self.API_USERS = self._get_api_users()
+
+    def __call__(self, request):
+
+        # Code to be executed for each request before
+        # the view (and later middleware) are called.
+
+        try:
```

Comment:
Would it have any performance effect if GET requests wouldn't get called _identify_api_caller at all, since it is not really necessary as GETs do not need to be authenticated? I suppose this way the code is cleaner if performance-wise there's no difference.

Reply:
There is practically no performance implication, since it is literally just a few lines of python code in case the header is left out. I wanted to include the possibility of using the header also during GET because I have a feeling the caller will be useful to know... Possibly for statistics gathering reasons, or optimizations (dont fetch all relations for some callers). A thin motivation perhaps, but the header can just be left out. Im going to clean it a little bit anyway because it looks odd.